### ISSUE TITLE
Full text searching support on the documents?

### ISSUE THREAD

**Issue by jeremydmiller** [JasperFx/marten#39]
More research on this one.

**Comment by jeffdoolittle** [comment]
Part 2 of of the @robconery series on postgres as document store may be relevant to this issue:

http://rob.conery.io/category/postgres-document-api/

**Comment by danbarua** [comment]
You could do this the same way you've done searchable fields, just persist it as a ```tsvector```.

**Comment by jeremydmiller** [comment]
@danbarua Yeah, but I haven't thought about how you might use or expose that from the client. You always have the ability to go straight at SQL I guess. Any suggestions on that angle?

**Comment by danbarua** [comment]
I'll have a play with it and see if I can come up with anything.

**Comment by danbarua** [comment]
Just leaving this here as a reminder:

```csharp
class Foo
{
  [FullText]
  public string Name {get; set;}
}

db.Query<Foo>.Where(x => x.Name.MatchesTextQuery("your_full_text_query"))
```

**Comment by robconery** [comment]
You have to persist the values at save - unfortunately triggers won't work on documents (which is a bummer). What we did for Moebius (Elixir db tool) was to define a function that would index a set of fields;

```ex
db(:users)
  |> searchable([:name, :email])
  |> save(...)
```

I would imagine here you could do the same with an attribute on the document itself:

```cs
[SearchableDocument("name","email")]
class Foo {
  //...
}
```

The when save is called check the attributes and update them in a separate call after save:

```sql
-- spin up this SQL string based on the attributes
update foos set(search) = to_tsvector(concat(name, ' ', email, ' '));
```

Anyway - that's what we do :).

**Comment by jeremydmiller** [comment]
@robconery Thanks for the input Rob. 

In Marten's internals, we've got a class called `DocumentMapping` that models everything about how to persist a document type (what's the id, which fields are duplicated for searching, etc), and from that model Marten codegen's the db table, upsert sproc, and even the C# glue code that goes from a document object to calling the right sproc.

In this case I think we'd add the full text searchable fields to the DocumentMapping model and enhance the codegen stuff to pass that along to upsert sprocs that also set up the full text mapping.

**Comment by robconery** [comment]
Sounds like it fits right in - nice job :).

**Comment by danbarua** [comment]
That's pretty much the approach I'm trying :+1:

**Comment by superlogical** [comment]
Here's another approach that uses Trigram's and the pg_trgm extension http://www.postgresql.org/docs/current/static/pgtrgm.html

Gitlab are using it to speed up like matches
https://about.gitlab.com/2016/03/18/fast-search-using-postgresql-trigram-indexes/

**Comment by phillip-haydon** [comment]
WIP - API Design:

User implements
```csharp
    public class Photo
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string[] Tags { get; set; }

        //...
    }

    public class PhotoSearch : Searchable<Photo>
    {
        public string ByNameAndDescription { get; set; }

        public string ByTag { get; set; }
    }
```


Marten implementation

```csharp
    public class Searchable<T> where T : class, new()
    {
        public IList<SearchResult<T>> Results { get; set; }

        public Func<SearchResult<T>, object> OrderBy { get; set; } = x => x.Rank;

        public Directions Direction { get; set; } = Directions.Descending;
    }

    public abstract class SearchResult<T>
    {
        public T Row { get; set; }

        public decimal Rank { get; set; }
    }

    public enum Directions
    {
        Ascending,
        Descending
    }
```


Schema configuration
```csharp
_.Schema.For<Photo>()
 .Searchable<PhotoSearch>(x =>
 {
    x.Vector(f => f.ByNameAndDescription)
     .With(f => f.Name, Weighting.High)
     .With(f => f.Description, Weighting.Low);
     
    x.Vector(f => f.ByTag)
     .With(f => f.Tags);
 });
```

Example query
```csharp
var result = session.Search<PhotoSearch>()
                    .Include<Member>(...)
                    .Against(x => x.ByNameAndDescription, term)
                    .And(x => x.ByTag, term);


var result = session.Search<PhotoSearch>()
                    .Include<Member>(...)
                    .Against(x => x.ByNameAndDescription, term)
                    .Or(x => x.ByTag, term)
                    .OrderBy(x => x.Rank, Directions.Ascending);
```

Example usage of result

```csharp
foreach (var photo in result)
{
    photo.Row.Name ...
    photo.Row.Tags ...
    
    photo.Rank // search result rank
}
```

**Comment by mdissel** [comment]
Nice! Can you still query on other non-fulltext fields in the Photo class (in your sample)?

**Comment by phillip-haydon** [comment]
Hmmm good point, I didn't consider that.

The problem I'm trying to solve is:


You can assign weighting against different terms as part of the `tsvector` index, but you can't specify a property in the `tsvector`

Meaning you cannot say: 

> Search against the Name field within the tsvector

So by creating a searchable doc we can have multiple tsvector columns for different search queries. The tsvectors can be rather small so this shouldn't be an issue to have a few extra columns.

But to have the API searchable I've tried to string it together. I guess it's now a case of exposing `T` so that we can also include normal criteria, something like:
```csharp
var result = session.Search<PhotoSearch>()
                    .Include<Member>(...)
                    .Where(x => x.Create > DateTime.UtcNow.AddDays(-10))
                    .Against(x => x.ByNameAndDescription, term)
                    .Or(x => x.ByTag, term)
                    .OrderBy(x => x.Rank, Directions.Ascending);
```

**Comment by mdissel** [comment]
Another option I'm missing is the stemming part. Where can we configure the stemming option that should be used (and the 'simple' stemming to disable stemming)

A nice to have is the ts_headline option ;)

**Comment by oskardudycz** [comment]
I'm closing this issue as it was solved with https://github.com/JasperFx/marten/pull/1098. Full text search is now available from Marten 3.2.0
